adding generic resource modules

zikalino opened the pull request.

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2018 Zim Kalinowski, <zikalino@microsoft.com>
```
isn't license under Microsoft?

zikalino replied:
this is the pattern from all the ansible modules, there's always original author + GNU license below

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+options:
+  url:
+    description:
+      - Azure RM Resource URL.
```
add more properties like required?

yungezz replied:
pls add properties like required or not. choices if there's and type.

zikalino replied:
if not required it should be omitted, only if required should be "required: yes".

zikalino replied:
also same applies to type, if type is string, then it's default and no need to include in the documentation.

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+  resource_name:
+    description:
+      - Resource name, should be specified if needed and URL is not specified
+  subresource_type:
```
subresource should be array

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+      - The HTTP method of the request or response. It MUST be uppercase.
+        choices: [ "GET", "PUT", "POST", "HEAD", "PATCH", "DELETE", "MERGE" ]
+        default: "GET"
+  provider:
```
duplicate

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+  resource_group:
+    description:
+      - Resource group to be used, should be specified if needed and URL is not specified
+  resource_type:
```
add choices?

zikalino replied:
there would be too many, maybe around 1000 and they depend on provider, so that would be too complex

yungezz replied:
or at least give user a link on where to find provider list. otherwise, it's hard for user to know what should be a right value for string type option.

zikalino replied:
yes, i think i can add reference to Azure API. Users of this module should be considered advanced, and they won't be able to write task from scratch without referring to REST API, using existing sample or using VSC extension

zikalino replied:
done

yungezz replied:
so no required options for this module?

zikalino replied:
only conditionally required, and conditions may be pretty complex, so this will be validated by underlying rest.

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+        self.mgmt_client = self.get_mgmt_svc_client(GenericRestClient,
+                                                    base_url=self._cloud_environment.endpoints.resource_manager)
+
+        if self.state == 'absent':
```
didn't get the point, when state='absent', then delete

zikalino replied:
that's how Ansible modules work:
- present (default) - resource should be created or updated
- absent - resource should be absent after module is run, meaning deleted.
my intention here is to make module as similar as possible to Ansible pattern (despite it's generic module).
So, by default, method will be PUT if present and DELETE if absent.
User may still use method parameter to override PUT with POST or anything else in some cases.

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+            self.url = self.url
+        else:
+            # URL is None, so we should construct URL from scratch
+            self.url = '/subscriptions/' + self.subscription_id
```
wrap line 200 - 215 as method for readability, eg. constructRestUri?

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+                self.url += '/resourcegroups/' + self.resource_group
+
+            if self.provider is not None:
+                self.url += '/providers/Microsoft.' + self.provider
```
provider check? to fail earlier

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,296 @@
+        super(GenericRestClientConfiguration, self).__init__(base_url)
+
+        self.add_user_agent('genericrestclient/1.0')
+        self.add_user_agent('Azure-SDK-For-Python')
```
how about add ansible user agent so that we could track usage from ansible?

yungezz commented on library/azure_rm_resource_facts.py:
```diff
@@ -0,0 +1,256 @@
+    description:
+      - Azure RM Resource URL.
+  provider:
+    description:
```
and more prorperties, like required or not, choices, default in all options

zikalino replied:
default is none for all options

yungezz commented on library/azure_rm_resource_facts.py:
```diff
@@ -0,0 +1,256 @@
+            self.url = self.url
+        else:
+            # URL is None, so we should construct URL from scratch
+            self.url = '/subscriptions/' + self.subscription_id
```
reusable code

yungezz commented on library/azure_rm_resource_facts.py:
```diff
@@ -0,0 +1,256 @@
+        return json.loads(response.text)
+
+
+class GenericRestClientConfiguration(AzureConfiguration):
```
reusable code

yungezz commented on tests/integration/targets/azure_rm_resource/tasks/main.yml:
```diff
@@ -0,0 +1,22 @@
+- name: Prepare random number
+  set_fact:
```
pls add test to cover subresource

zikalino replied:
ok

zikalino replied:
i didn't want to make any changes to azure_rm_common.py. but now i think it should be a good idea to have azure_rm_common_rest or sth like that

zikalino replied:
i created another common file, so didn't have to touch azure_rm_common at this point

yungezz replied:
how about name it as azure_rm_rest_common.py

zikalino commented:
@yungezz 
i have made major updates to your module according to your comments
- moved common code to a separate file
- use built in rest api function to form URL
- changed API for subresources, so now a list can be specified (matching REST API URL function)
- updated documentation to refer Azure REST API docs
- updated test so now it includes subresource testing and also absent / delete test
- updated test to include facts queries

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,234 @@
+            rargs = dict()
+            rargs['subscription'] = self.subscription_id
+            rargs['resource_group'] = self.resource_group
+            if not (self.provider is None or self.provider.lower().startswith('.microsoft')):
```
is rest url case sensitive?

zikalino replied:
URLs are never case sensitive.

yungezz replied:
http://www.w3.org/TR/WD-html40-970708/htmlweb.html

zikalino replied:
yes, that's what the spec says , however, nowadays nobody makes urls case-sensitive. Same with Azure REST API. It's not case sensitive.

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,234 @@
+            rargs['subscription'] = self.subscription_id
+            rargs['resource_group'] = self.resource_group
+            if not (self.provider is None or self.provider.lower().startswith('.microsoft')):
+                rargs['namespace'] = "microsoft." + self.provider
```
any validation on provider?

zikalino replied:
fixed one check (if None)
basic validation will be done by underlying azure rest module.
If someone specifies wrong provider, whole request will fail.

yungezz commented on library/azure_rm_resource.py:
```diff
@@ -0,0 +1,234 @@
+'''
+
+EXAMPLES = '''
+  - name: Update scaleset info using azure_rm_resource
```
do you want to add some sentences or samples about using this module could provide azure resources which hasn't delegated module yet?

zikalino replied:
maybe a good idea
